examples/parser/Parser/Core.roc

```diff
@@ -0,0 +1,292 @@
+fail : Str -> Parser * *
+fail = \msg ->
+    buildPrimitiveParser \_input -> Err (ParsingFailure msg)
+
+## Parser that will always produce the given `val`, without looking at the actual input.
+## This is useful as basic building block, especially in combination with
+## `map` and `apply`.
+const : a -> Parser * a
+const = \val ->
+    buildPrimitiveParser \input ->
+        Ok { val: val, input: input }
+
+## Try the `first` parser and (only) if it fails, try the `second` parser as fallback.
+alt : Parser input a, Parser input a -> Parser input a
+alt = \first, second ->
+    buildPrimitiveParser \input ->
+        when parsePartial first input is
+            Ok { val: val, input: rest } -> Ok { val: val, input: rest }
+            Err (ParsingFailure firstErr) ->
+                when parsePartial second input is
+                    Ok { val: val, input: rest } -> Ok { val: val, input: rest }
+                    Err (ParsingFailure secondErr) ->
+                        Err (ParsingFailure ("\(firstErr) or \(secondErr)"))
+
+## Runs a parser building a function, then a parser building a value,
+## and finally returns the result of calling the function with the value.
+##
+## This is useful if you are building up a structure that requires more parameters
+## than there are variants of `map`, `map2`, `map3` etc. for.
+##
+## For instance, the following two are the same:
+##
+## >>> const (\x, y, z -> Triple x y z)
+## >>> |> map3 Parser.Str.nat Parser.Str.nat Parser.Str.nat
+##
+## >>> const (\x -> \y -> \z -> Triple x y z)
+## >>> |> apply Parser.Str.nat
+## >>> |> apply Parser.Str.nat
+## >>> |> apply Parser.Str.nat
+##
+## (And indeed, this is how `map`, `map2`, `map3` etc. are implemented under the hood.)
+##
+## # Currying
+## Be aware that when using `apply`, you need to explicitly 'curry' the parameters to the construction function.
+## This means that instead of writing `\x, y, z -> ...`
+## you'll need to write `\x -> \y -> \z -> ...`.
+## This is because the parameters to the function will be applied one-by-one as parsing continues.
+apply : Parser input (a -> b), Parser input a -> Parser input b
+apply = \funParser, valParser ->
+    combined = \input ->
+        { val: funVal, input: rest } <- Result.try (parsePartial funParser input)
+        parsePartial valParser rest
+        |> Result.map \{ val: val, input: rest2 } ->
+            { val: funVal val, input: rest2 }
+
+    buildPrimitiveParser combined
+
+# Internal utility function. Not exposed to users, since usage is discouraged!
+#
+# Runs `firstParser` and (only) if it succeeds,
+# runs the function `buildNextParser` on its result value.
+# This function returns a new parser, which is finally run.
+#
+# `andThen` is usually more flexible than necessary, and less efficient
+# than using `const` with `map` and/or `apply`.
+# Consider using those functions first.
+andThen : Parser input a, (a -> Parser input b) -> Parser input b
+andThen = \firstParser, buildNextParser ->
```

Comment:
It seems like most usages of `andThen` can be achieved with a combination of `alt` and `apply`, no? Are there instances where that is not the case?

Reply:
Correct. Any context-free language is parseable using only `alt`, `apply` and `map`. They are together just as powerful as one can do with a parser generator like Yacc and co. 
The reason why many combinator-based parsers define `andThen` is twofold:
- `andThen` is monadic bind, so systems that have special support for monadic notation can use its syntactic sugar. This is also true in Roc where we can use backpassing with `andThen`.
- It is easy to define `alt`, `apply`, `map` in terms of `andThen`. But this comes with two big caveats:

1. It is more difficult to optimize
2. The rest of the parser is only made ready at runtime. Thus, `andThen` is a black box. You cannot look at the structure of a parser that uses `andThen` without running it on actual input.

As such, there is much to be said for not supporting it at all, and instead keeping with applicative parsing only. Some cool stuff you can do if you are able to inspect the parser without running it:
- build a CLI options parser that can print the help text of supported options automatically.
- bidirectional parsing: One definition that can be used for both parsing and prettyprinting.
- More clever optimizations before running the parser by rewriting the parser's AST to an equivalent but faster format. This is especially relevant when e.g. building a parser for regular languages because there there is a LOT of optimization possible by rewriting the state machine.